src/theory/datatypes/type_enumerator.cpp

```diff
@@ -166,59 +192,125 @@ Node DatatypesEnumerator::getTermEnum( TypeNode tn, unsigned i ){
-     d_sel_index.push_back( std::vector< unsigned >() );
-     d_sel_sum.push_back( -1 );
-   }else{
-     // find the "zero" constructor via mkGroundTerm
+     d_sel_types.push_back(std::vector<TypeNode>());
+     d_sel_index.push_back(std::vector<unsigned>());
+     d_sel_sum.push_back(-1);
+   }
+   else
+   {
+     // find the "zero" term via mkGroundTerm
      Debug("dt-enum-debug") << "make ground term..." << std::endl;
-     Node t = d_type.mkGroundTerm();
-     Debug("dt-enum-debug") << "done : " << t << std::endl;
-     Assert(t.getKind() == kind::APPLY_CONSTRUCTOR);
-     // start with the constructor for which a ground term is constructed
-     d_zeroCtor = datatypes::utils::indexOf(t.getOperator());
+     // Start with the ground term constructed via mkGroundTerm, which does
+     // a traversal over the structure of the datatype to find a finite term.
+     d_zeroTerm = Node::fromExpr(d_datatype.mkGroundValue(d_type.toType()));
+     d_zeroTermActive = true;
+     Debug("dt-enum-debug") << "done : " << d_zeroTerm << std::endl;
+     Assert(d_zeroTerm.getKind() == kind::APPLY_CONSTRUCTOR);
      d_has_debruijn = 0;
    }
-   Debug("dt-enum") << "zero ctor : " << d_zeroCtor << std::endl;
-   d_ctor = d_zeroCtor;
-   for( unsigned i=0; i<d_datatype.getNumConstructors(); ++i ){
-     d_sel_types.push_back( std::vector< TypeNode >() );
-     d_sel_index.push_back( std::vector< unsigned >() );
-     d_sel_sum.push_back( -1 );
+   Debug("dt-enum") << "zero term : " << d_zeroTerm << std::endl;
+   d_ctor = 0;
+   for (unsigned i = 0, ncons = d_datatype.getNumConstructors(); i < ncons; ++i)
+   {
+     d_sel_types.push_back(std::vector<TypeNode>());
+     d_sel_index.push_back(std::vector<unsigned>());
+     d_sel_sum.push_back(-1);
      DatatypeConstructor ctor = d_datatype[i];
      Type typ;
-     if( d_datatype.isParametric() ){
+     if (d_datatype.isParametric())
+     {
        typ = ctor.getSpecializedConstructorType(d_type.toType());
      }
-     for( unsigned a = 0; a < ctor.getNumArgs(); ++a ){
+     for (unsigned a = 0; a < ctor.getNumArgs(); ++a)
+     {
        TypeNode tn;
-       if( d_datatype.isParametric() ){
-         tn = TypeNode::fromType( typ )[a];
-       }else{
+       if (d_datatype.isParametric())
+       {
+         tn = TypeNode::fromType(typ)[a];
+       }
+       else
+       {
          tn = Node::fromExpr(ctor[a].getSelector()).getType()[1];
        }
-       d_sel_types.back().push_back( tn );
-       d_sel_index.back().push_back( 0 );
+       d_sel_types.back().push_back(tn);
+       d_sel_index.back().push_back(0);
      }
-     if( !d_sel_index.back().empty() ){
+     if (!d_sel_index.back().empty())
+     {
        d_sel_index.back().pop_back();
      }
    }
    d_size_limit = 0;
-   //set up initial conditions (should always succeed)
-   ++*this; //increment( d_ctor );
+   if (!d_zeroTermActive)
+   {
+     // set up initial conditions (should always succeed)
+     ++*this;
+   }
```

Comment:
What is happening here?

Reply:
The process for constructing the next term in the enumeration is being called. This is necessary to do since setting up the first term is analogous to setting up each subsequent term.